rainier-core/src/main/scala/com/stripe/rainier/core/RandomVariable.scala

```diff
@@ -226,4 +238,85 @@ object RandomVariable {
   def fit[L, T](l: L, seq: Seq[T])(
       implicit toLH: ToLikelihood[L, T]): RandomVariable[Unit] =
     toLH(l).fit(seq)
+
+  private[core] def waic(samples: List[Array[Double]],
+                         variables: List[Variable],
+                         target: Target): Double =
+    if (target.nRows == 0)
+      0.0
+    else {
+      val cf = Compiler.default.compile(
+        target.placeholderVariables ++ variables,
+        List(("logLikelihood", target.real)))
+      val globalBuf = new Array[Double](cf.numGlobals)
+      val inputBuf = new Array[Double](cf.numInputs)
+      val data = target.placeholderVariables.map { v =>
+        target.placeholders(v)
+      }.toArray
+      val sampleArray = samples.toArray
+      var i = 0
+      var result = 0.0
+      while (i < target.nRows) {
+        result += waic(sampleArray, cf, inputBuf, globalBuf, data, i)
+        i += 1
+      }
+      result * -2.0
+    }
+
+  private def waic(samples: Array[Array[Double]],
+                   cf: CompiledFunction,
+                   inputBuf: Array[Double],
+                   globalsBuf: Array[Double],
+                   data: Array[Array[Double]],
+                   i: Int): Double = {
+    val nInputs = inputBuf.size
+    val nDataInputs = data.size
+    val nSampleInputs = nInputs - nDataInputs
+    val nSamples = samples.size
+    val outputs = new Array[Double](nSamples)
+    var j = 0
+    while (j < nDataInputs) {
+      inputBuf(j) = data(j)(i)
+      j += 1
+    }
+    var k = 0
+    while (k < nSamples) {
+      val sample = samples(k)
+      j = 0
+      while (j < nSampleInputs) {
+        inputBuf(nDataInputs + j) = sample(j)
+        j += 1
+      }
+      outputs(i) = cf.output(inputBuf, globalsBuf, 0)
+      k += 1
+    }
+
+    val max = outputs.max
+    var expSum = 0.0
+    k = 0
+    while (k < nSamples) {
+      expSum += Math.exp(outputs(i) - max)
+      k += 1
+    }
+    val lppd = Math.log(expSum) + max - Math.log(nSamples.toDouble)
```

Comment:
Ahh, took me a few minutes to realize that you moved the division for the average into a subtraction in log space. Maybe worth a comment.